src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,340 @@
+                            statementsCannotBeConverted = new[] { current };
+                        }
+
+                        break;
+
+                    case SyntaxKind.EmptyStatement:
+                        // The innermost statement is an empty statement, stop processing
+                        // Example:
+                        // foreach(...)
+                        // {
+                        //    ;<- empty statement
+                        // }
+                        statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        break;
+
+                    default:
+                        // If no specific case found, stop processing.
+                        statementsCannotBeConverted = new[] { current };
+                        break;
+                }
+            }
+
+            // Trailing tokens are collected in the reverse order: from extrenal block down to internal ones. Reverse them.
+            trailingTokensBuilder.Reverse();
+
+            return new ForEachInfo<ForEachStatementSyntax, StatementSyntax>(
+                forEachStatement,
+                semanticModel,
+                convertingNodesBuilder.ToImmutable(),
+                identifiersBuilder.ToImmutable(),
+                statementsCannotBeConverted.ToImmutableArray(),
+                currentLeadingTokens.ToImmutableArray(),
+                trailingTokensBuilder.ToImmutable());
+
+            void ProcessLocalDeclarationStatement(LocalDeclarationStatementSyntax localDeclarationStatement)
+            {
+                var localDeclarationLeadingTrivia = new IEnumerable<SyntaxTrivia>[] {
+                    currentLeadingTokens.GetTrivia(),
+                    localDeclarationStatement.Declaration.Type.GetLeadingTrivia(),
+                    localDeclarationStatement.Declaration.Type.GetTrailingTrivia() }.Flatten();
+                var localDeclarationTrailingTrivia = SyntaxNodeOrTokenExtensions.GetTrivia(localDeclarationStatement.SemicolonToken);
+                var separators = localDeclarationStatement.Declaration.Variables.GetSeparators().ToArray();
+                for (var i = 0; i < localDeclarationStatement.Declaration.Variables.Count; i++)
+                {
+                    var variable = localDeclarationStatement.Declaration.Variables[i];
+                    convertingNodesBuilder.Add(new ExtendedSyntaxNode(
+                        variable,
+                        i == 0 ? localDeclarationLeadingTrivia : separators[i - 1].TrailingTrivia,
+                        i == localDeclarationStatement.Declaration.Variables.Count - 1
+                            ? (IEnumerable<SyntaxTrivia>)localDeclarationTrailingTrivia
+                            : separators[i].LeadingTrivia));
+                    identifiersBuilder.Add(variable.Identifier);
+                }
+
+                currentLeadingTokens = new List<SyntaxToken>();
```

Comment:
>currentLeadingTokens [](start = 16, length = 20)

Can we move this statement closer to the place where old `currentLeadingTokens` are used and are no longer required?